### PR TITLE
New metadata format

### PR DESCRIPTION
**Pull request by kaczmarczyck**
This PR changes to metadata format to include the signature, and a version instead of a timestamp.

tl;dr tell me how to shuffle the format around:
- 32 B upgrade hash (SHA256)
- 64 B signature,
- 32 B padding,
- 20 B version (8 byte used) and
-  4 B partition address in little endian encoding.

There are a few things we can discuss about the format:

1. The version is currently `u64`, since the GetInfo command returns an integer, and our CBOR parser works with up to 63 bit integers. We could make it more sophisticated like SEMVER, but would have to convert it to 63 bit eventually.
2. More padding before the signed part of the metadata starts. Would allow us to write different / bigger / PQC signatures there. I wanted to get your opinions before deciding on a number.

Some things are missing:
- Versions are only checked in the bootloader, but not yet in CTAP. This is consistent with existing behavior, but will probably be stricter soon.
- Code has not moved to TockEnv yet. The command will stay in CTAP, but the Metadata logic will end up there.

### CODE REVIEW

File: src/ctap/mod.rs [hunk 1]
```diff
@@ -212,45 +211,55 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> &str {
...
+    public_key_bytes: &[u8],
     metadata: &[u8],
-) -> Result<[u8; 32], Ctap2StatusCode> {
-    const METADATA_LEN: usize = 40;
+) -> Result<(), Ctap2StatusCode> {
+    const METADATA_LEN: usize = 0x1000;
     if metadata.len() != METADATA_LEN {
         return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
     }
+    let metadata_address = LittleEndian::read_u32(&metadata[148..][..4]);
```
Comment (ia0) [inline src/ctap/mod.rs]: Shouldn't we ideally use constants for offsets in the metadata? (0, 32, 140, 148)

File: src/ctap/mod.rs [hunk 2]
```diff
@@ -212,45 +211,55 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> &str {
...
+    if metadata_address as usize != upgrade_locations.partition_address() {
+        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+    }
     // The hash implementation handles this in chunks, so no memory issues.
     let partition_slice = upgrade_locations
         .read_partition(0, upgrade_locations.partition_length())
         .map_err(|_| Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR)?;
     let mut hasher = Sha256::new();
+    hasher.update(&metadata[128..METADATA_LEN]);
```
Comment (ia0) [inline src/ctap/mod.rs]: I didn't see where it is documented that the hash starts at offset 128.

**Review by ia0 — DISMISSED** [review]
I didn't look too much in the Python logic.

Regarding how much non-signed padding, I guess it depends how much bigger checksums and signatures can be in the future. I don't know but looks like the metadata allocated size is quite huge (4096 bytes), so there's room. Why not just split in 2 parts? [0-2048) non-signed, and [2048-4096) signed?

File: bootloader/src/main.rs [hunk 3]
```diff
@@ -77,14 +75,14 @@ impl BootPartition {
     const FIRMWARE_LENGTH: usize = 0x00040000;
 
     /// Reads the metadata, returns the timestamp if all checks pass.
```
Comment (zhalvorsen) [inline bootloader/src/main.rs]:
```suggestion
    /// Reads the metadata, returns the firmware version if all checks pass.
```

**Review by zhalvorsen — DISMISSED** [review]

File: src/ctap/mod.rs (hunk 2 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: I added a constant and updated all documentation.

File: src/ctap/mod.rs (hunk 1 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: I kept some numbers that only matter inside the function that uses them. I can also make them const, if you like. I went with your suggestion to use half pages.

File: src/ctap/mod.rs [hunk 4]
```diff
@@ -211,37 +211,43 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> &str {
 
 /// Parses the metadata of an upgrade, and checks its correctness.
 ///
-/// The metadata consists of:
+/// The metadata is a page starting with:
```
Comment (ia0) [inline src/ctap/mod.rs:214]: By the way, this is specific to Nordic. A page is not necessarily 4kB. It makes sense for the bootloader which is Nordic-specific, but ctap-code should ideally be abstract over a page size. But I guess it's fine to keep it like that or rename "page" to "slice" or something.

**Review by ia0 — APPROVED** [review]

File: src/ctap/mod.rs (hunk 4 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs:214]: Ah yeah, no problem. This is moving to Env in 2 PRs.
